include/dxc/Support/Path.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include "llvm/ADT/StringRef.h"
+#include <string>
+
+namespace hlsl {
+
+template <typename CharTy>
+bool IsAbsoluteOrCurDirRelativeShared(const CharTy *Path) {
+  if (!Path || !Path[0])
+    return false;
+  // Current dir-relative path.
+  if (Path[0] == '.') {
+    return Path[1] == '\0' || Path[1] == '/' || Path[1] == '\\';
+  }
+  // Disk designator, then absolute path.
+  if (Path[1] == ':' && (Path[2] == '\\' || Path[2] == '/')) {
+    return true;
+  }
+  // UNC name
+  if (Path[0] == '\\') {
+    return Path[1] == '\\';
+  }
+
+#ifndef _WIN32
+  // Absolute paths on unix systems start with '/'
+  if (Path[0] == '/') {
+    return true;
+  }
+#endif
+
+  //
+  // NOTE: there are a number of cases we don't handle, as they don't play well
+  // with the simple file system abstraction we use:
+  // - current directory on disk designator (eg, D:file.ext), requires per-disk
+  // current dir
+  // - parent paths relative to current directory (eg, ..\\file.ext)
+  //
+  // The current-directory support is available to help in-memory handlers.
+  // On-disk handlers will typically have absolute paths to begin with.
+  //
+  return false;
+}
+
+inline bool IsAbsoluteOrCurDirRelativeW(const wchar_t *Path) {
+  return IsAbsoluteOrCurDirRelativeShared<wchar_t>(Path);
+}
+inline bool IsAbsoluteOrCurDirRelative(const char *Path) {
+  return IsAbsoluteOrCurDirRelativeShared<char>(Path);
+}
+
+template <typename CharT, typename StringTy>
+inline StringTy NormalizePathImpl(const CharT *Path, size_t Length,
+                                  bool PrefixWithDot) {
+  StringTy PathCopy(Path, Length);
+
+#ifdef _WIN32
```

Comment:
So this means that I'll get different looking paths if compiled on windows vs non-windows, even if all of my paths are relative.  Will this cause any problems with debuggers that only run on one platform?

Reply:
That's right. The main reason I'm doing this is because of UNC paths on windows. If we normalize all paths to forward slash, UNC paths stop working. If we just leave UNC paths unchanged just for windows, we'll get different looking paths anyway.

I don't think debuggers are going to have any problems. All the paths they get will go through the same normalization process.